Objects/complexobject.c

```diff
@@ -905,6 +905,67 @@ complex_subtype_from_string(PyTypeObject *type, PyObject *v)
     return result;
 }
 
+static PyObject *
+actual_complex_new(PyTypeObject *type, PyObject *args, PyObject *kwargs)
+{
+    PyObject *res = NULL;
+    PyNumberMethods *nbr;
+
+    if (PyTuple_GET_SIZE(args) > 1 || (kwargs != NULL && PyDict_GET_SIZE(kwargs))) {
+        return complex_new(type, args, kwargs);
+    }
+    if (!PyTuple_GET_SIZE(args)) {
+        return complex_subtype_from_doubles(type, 0, 0);
+    }
+
+    PyObject *arg = PyTuple_GET_ITEM(args, 0);
+    /* Special-case for a single argument when type(arg) is complex. */
+    if (PyComplex_CheckExact(arg) && type == &PyComplex_Type) {
+        /* Note that we can't know whether it's safe to return
+           a complex *subclass* instance as-is, hence the restriction
+           to exact complexes here.  If either the input or the
+           output is a complex subclass, it will be handled below
+           as a non-orthogonal vector.  */
+        return Py_NewRef(arg);
+    }
+    if (PyUnicode_Check(arg)) {
+        return complex_subtype_from_string(type, arg);
+    }
+    PyObject *tmp = try_complex_special_method(arg);
+    if (tmp) {
+        Py_complex c = ((PyComplexObject*)tmp)->cval;
+        res = complex_subtype_from_doubles(type, c.real, c.imag);
+        Py_DECREF(tmp);
+    }
+    else if (PyErr_Occurred()) {
+        return NULL;
+    }
+    else if (PyComplex_Check(arg)) {
```

Comment:
FYI, this is not tested.  But I'm not sure if that's a right logic.  Complex subclasses should be covered by ``try_complex_special_method()``, c.f. PyNumber_Float().

Reply:
`try_complex_special_method` is relatively expensive in comparison to `PyNumber_Float()`, because there is no `nb_complex` slot. And `__complex__` is looked up even for exact `complex`, `float` and `int`. I planned to do something with this, but in a different PR. We can not also completely exclude `complex` subclasses without `__complex__`.

Reply:
> try_complex_special_method is relatively expensive in comparison to PyNumber_Float()

That seems to be an implementation detail.  I wonder if we could add nb_complex slot: there is a reserved slot right now anyway.

> And __complex__ is looked up even for exact complex, float and int.

The current (i.e. in the main) code - uses here same logic as the float constructor: there is a case for exact complex (as for exact float in PyNumber_Float()).

> We can not also completely exclude complex subclasses without __complex__.

People could break thing in very crazy ways, but should we support such cases?  (Looks as a variant of #112636.)

Reply:
FYI, discourse thread: https://discuss.python.org/t/use-nb-reserved-slot-of-pynumbermethods-to-support-complex-method/56620